### PR TITLE
feat: auto-chunk oversize submissions in enqueue/enqueue_batch

### PR DESCRIPTION
**Pull request by drake-nominal**
## What this fixes

Two related issues in `NominalDatasetStream`:

1. **Oversize-submission silent rejection.** `enqueue` and `enqueue_batch` previously accepted submissions of any size but produced one `WriteRequest` containing all points — submissions exceeding `max_points_per_record` (default 250k) could result in a request that the backend would reject.
2. **TOCTOU race in capacity enforcement.** `SeriesBuffer::has_capacity` was checked outside the lock, so two concurrent producers could both see capacity available on an empty buffer, both acquire the lock sequentially, and write `2*cap` points — producing oversize `WriteRequest`s under multi-producer Rust callers.

## What changed

- New private chunking helpers in `stream.rs`:
  - `chunk_points(points, cap)` splits one `PointsType` into ≤cap-sized pieces across all 8 match arms.
  - `chunk_by_capacity(batch, cap)` pre-splits oversize entries and greedy-packs into rounds where each round's total ≤ cap.
- `enqueue` and `enqueue_batch` rewritten to loop over chunks, each going through `when_capacity` independently (per-chunk backpressure).
- `SeriesBuffer::has_capacity` + `lock` pair replaced by `lock_if_capacity(n)` which takes the lock and checks `count + n ≤ max_capacity` atomically under the lock; returns `None` (releasing the lock) if not. `on_notify` replaced by `wait_for_capacity` which re-checks the invariant in a loop on every wakeup. `notify` switched to `notify_all` so multiple waiters can each re-check.
- Doc updates on `max_points_per_record`, `enqueue`, `enqueue_batch`, and the new buffer primitives.

A submission of N points now produces `ceil(N / max_points_per_record)` right-sized requests instead of one oversize request that fails. Memory bounded by `max_capacity` at all times, including under concurrent producers. No public API change.

## How it was tested

Most tangibly, I called `enqueue_batch` with a 100 million entry array, and it successfully streamed to Nominal fairly performantly, but with a fraction of the thinking.

- 11 new unit + integration tests covering: `chunk_points` across all 8 match arms, `chunk_by_capacity` greedy-pack with edge cases (empty entries, exact-cap boundary, tags-survive-split), `enqueue` and `enqueue_batch` chunking under multiple shapes, intra-request ordering preservation, the `lock_if_capacity` invariant, a backpressure regression guard, and a multi-producer concurrency test (4 producers × 8 chunks each).
- The concurrency test deliberately reproduces the TOCTOU race: it **fails 50/50 against the previous `has_capacity`+`lock` pattern** and **passes 50/50 against `lock_if_capacity`**.
- All 19 stream tests pass; the chunking-related ones (`stream::tests::*`) run 25× under parallel load with zero flakes.
- `cargo fmt --all -- --check` clean. `cargo clippy --lib --tests -- -D warnings` clean.

### CODE REVIEW

File: nominal-streaming/src/stream.rs [hunk 1]
```diff
@@ -49,6 +56,11 @@ use crate::types::IntoTimestamp;
 
 #[derive(Debug, Clone)]
 pub struct NominalStreamOpts {
+    /// Upper bound on the number of points in any single downstream
+    /// `WriteRequest`. Producer submissions exceeding this are split internally
+    /// (see `NominalDatasetStream::enqueue` and `enqueue_batch`); the buffer
+    /// `SeriesBuffer::lock_if_capacity` enforces the bound atomically so the
+    /// invariant holds under concurrent producers.
```
Comment (alxhill) [inline nominal-streaming/src/stream.rs:63]: ugh llm comments love to explain the current PR like it's the most important part of the code 🤦

File: nominal-streaming/src/stream.rs [hunk 2]
```diff
@@ -378,40 +390,95 @@ impl NominalDatasetStream {
...
-                &self.primary_buffer
-            } else {
-                info!("waiting for secondary buffer to flush to append {new_count} points...");
-                self.secondary_handle.thread().unpark();
-                &self.secondary_buffer
-            };
-
-            buf.on_notify(callback);
+            callback(guard);
+            return;
```
Comment (alxhill) [inline nominal-streaming/src/stream.rs]: I believe this is unnecessary - NominalDatasetStream is not Sync so there's no TOCTOU race (will confirm)

Comment (alxhill) [inline nominal-streaming/src/stream.rs:465]: hmm I thought NominalDatasetStream couldn't be shared between threads, but this method isn't &mut so it could be - sad. The other option here is to keep this code the same and make the method &mut - more of a break, but also simpler / doesn't change this (battle tested on one thread) part of the code

Comment (alxhill) [reply nominal-streaming/src/stream.rs:465]: okay did some more research here - tl;dr while this is technically a TOCTOU issue, the practical implication is just that the max buffer size is `max_capacity * num_threads`, which while not ideal is also not the end of the world (i.e it just affects the accuracy of the backpressure mechanism). making changes to the buffer / count tracking is fairly risky & the current code is working reliably, so I think we should leave it as-is.

(tried switching to `&mut self` instead of `&self`, but that makes it impossible to have more than one writer at a time - seems too restrictive)

Comment (joseph-gio) [reply nominal-streaming/src/stream.rs:465]: We have a read/write timeout set to two seconds, which means a sufficiently large batch size can likely time out

File: nominal-streaming/src/stream.rs (hunk 1 above)
Comment (drake-nominal) [reply nominal-streaming/src/stream.rs]: Removed the old PR-specific commentary. The option now documents the output point limit in two lines; the producer-side locking explanation is gone.

File: nominal-streaming/src/stream.rs (hunk 2 above)
Comment (drake-nominal) [reply nominal-streaming/src/stream.rs]: Addressed in 1f020bf. Buffer admission, count tracking, locking, and wake-ups are unchanged from main, and the API remains `&self`. Splitting now happens in the background processor after it takes the buffer contents and releases the lock. This caps output requests even when an oversized input or concurrent producers overfill a buffer, without tightening producer backpressure.

The oversized-request timeout concern remains the motivation for this fix. The point limit prevents one huge submission from becoming one huge request; it does not impose a byte limit or guarantee completion within the timeout. A regression also verifies that oversized enqueue returns while downstream consumption is blocked.

File: nominal-streaming/src/stream.rs [hunk 3]
```diff
@@ -1112,6 +1119,73 @@ fn request_dispatcher<C: WriteRequestConsumer + 'static>(
...
+    if count <= cap {
+        send(series, count);
+        return;
+    }
+    let mut record = Vec::new();
+    let mut count = 0;
+    for mut series in series {
+        let points = series.points.take().unwrap().points_type.unwrap();
+        for_each_points_chunk(points, cap, |points| {
+            let n = points_len(&points);
```
Comment (alxhill) [inline nominal-streaming/src/stream.rs]: nit: `for_each_points_chunk` calls `points_len` already, can pass in instead of recomputing

**Review by alxhill — APPROVED** [review]
lgtm! I added two follow up PRs testing some edge cases, up to you on how to merge them

File: nominal-streaming/src/stream.rs (hunk 3 above)
Comment (drake-nominal) [reply nominal-streaming/src/stream.rs]: Fixed in 000a766: the chunk callback now receives the already-known length, and the chunk regression checks that it matches the actual contents. Also incorporated the shutdown and saturation tests from #359 and #360, preserving your commits.
